### PR TITLE
Refact of storage related classes

### PR DESCRIPTION
**Pull request by emulgeator**

### CODE REVIEW

File: src/Storage/FileStorage.php [hunk 1]
```diff
@@ -1,476 +1,365 @@
...
+            $path .= DIRECTORY_SEPARATOR;
+        }
+
+        $this->path = $path;
+
+        $this->initAndValidatePath();
+    }
+
+    /**
+     * @throws StorageException
```
Comment (szeber) [inline src/Storage/FileStorage.php]: Would be nice to make specific exceptions to the different error types.

File: src/Storage/FileStorage.php [hunk 2]
```diff
@@ -1,476 +1,365 @@
...
+
+            $this->fileHandler->write($fullPath, $this->prepareDataToStore($data, $ttlInSeconds));
+
+            $item->setFinished();
+            $this->getDebugDataHandlerRegistry()->addStorage($item);
+        } catch (FileException $e) {
+            throw new StorageException('Unable to write data to FileStorage (file: ' . $fullPath . ' )', 0, $e);
+        }
+
+        // Disable potential warnings if unit testing with vfsStream
```
Comment (szeber) [inline src/Storage/FileStorage.php]: BTW we should probably replace vfsStream too... Flysystem may be a good option for file handling, not sure if it covers everything we need it for though...

File: src/Storage/FileStorage.php [hunk 3]
```diff
@@ -1,476 +1,365 @@
...
+        // If the storePlainText option is set to false, we support TTL functionality.
+        return !$this->canOnlyStorePlainText;
+    }
+
+    public function isReadOnly(): bool
+    {
+        return $this->readOnly;
+    }
+
+    public function isCanOnlyStorePlainText(): bool
```
Comment (szeber) [inline src/Storage/FileStorage.php]: The is prefix is completely unnecessary, canOnlyStorePlainText() implies it's a boolean as well

File: src/Storage/FileStorage.php [hunk 4]
```diff
@@ -1,476 +1,365 @@
...
+        return $this->fileModeOctal;
+    }
+
+    public function setFileModeOctal(int $fileModeOctal): self
+    {
+        $this->fileModeOctal = $fileModeOctal;
+        return $this;
+    }
+
+    public function isHashKey(): bool
```
Comment (szeber) [inline src/Storage/FileStorage.php]: isKeyHashed?

File: src/Storage/FileStorage.php [hunk 5]
```diff
@@ -1,476 +1,365 @@
...
+        $this->hashKey = $hashKey;
+        return $this;
+    }
+
+    public function getFileHandler(): FileHandlerPhp
+    {
+        return $this->fileHandler;
+    }
+
+    protected function getDateHelper(): DateHelper
```
Comment (szeber) [inline src/Storage/FileStorage.php]: Would be nice if this was injected as well

File: src/Database/DbConnection.php [hunk 6]
```diff
@@ -25,376 +25,399 @@
...
+     *
+     * @throws \YapepBase\Exception\DatabaseException   On execution errors.
+     */
+    public function query($query, array $params = [])
+    {
+        if (empty($this->connection)) {
+            throw new DatabaseException('Connection to the database is not established');
+        }
+
+        $backtrace    = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS,2);
```
[the remark this answers is not in the document]
Comment (emulgeator) [reply src/Database/DbConnection.php]: This will be needed by the debugger. This way we will know from where did the query come from. What I didn't connect yet. 
I was testing it to see if it works and fast and left the working version here as I'll need it later as well.
In the storage branch I refactored the Debugger as well because most of the debugger usage is in the storage, and I refact and test the Dao and DB related stuffs as well here (Not under storage namespace but they are related to the storage too)

File: src/Debug/DataHandlerRegistry.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        /** @var DateHelper $dateHelper */
+        $dateHelper = Application::getInstance()->getDiContainer()->get(DateHelper::class);
+
+        $this->initiatedAt = $dateHelper->getCurrentTimestampMs();
+    }
+
+    public function register(string $name, ICanReturnItems $dataHandler): void
+    {
+        if (isset($this->dataHandlers[$name])) {
```
[the remark this answers is not in the document]
Comment (emulgeator) [reply src/Debug/DataHandlerRegistry.php]: I was thinking about the same but decided that I give the possibility to register the same type twice if needed. The register can be still used by class name as a name though

File: src/Debug/DataHandlerRegistry.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    /** @var float */
+    protected $initiatedAt;
+
+    /** @var ICanReturnItems[] */
+    protected $dataHandlers = [];
+
+    public function __construct()
+    {
+        /** @var DateHelper $dateHelper */
+        $dateHelper = Application::getInstance()->getDiContainer()->get(DateHelper::class);
```
[the remark this answers is not in the document]
Comment (emulgeator) [reply src/Debug/DataHandlerRegistry.php]: Let's talk about this. We will need a general solution to follow everywhere

File: src/Debug/DataHandlerRegistry.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    {
+        $this->dataHandlers = [];
+    }
+
+    public function getInitiatedAt(): float
+    {
+        return $this->initiatedAt;
+    }
+
+    public function addCurlRequest(CurlRequest $item): void
```
[the remark this answers is not in the document]
Comment (emulgeator) [reply src/Debug/DataHandlerRegistry.php]: I wanted to have specific items for the ones handled by the FW.
But I add a general item type as well so from the application anything could be added

File: src/Storage/FileStorage.php (hunk 1 above)
Comment (emulgeator) [reply src/Storage/FileStorage.php]: yeah will need a lot more exceptions but you started the exception refact in the routing for some reason and don't want to conflict with anything there as it's not merged.
So at the end we'll need to go  through and check the exceptions

File: src/Storage/FileStorage.php (hunk 2 above)
Comment (emulgeator) [reply src/Storage/FileStorage.php]: To be honest I left this here because I don't know yet why would vfs trigger warnings.
Let's keep this flysystem thing to the end, when we have a workinf FW we can start experiment with these.
Though I'm almost sure it provides everything we need

File: src/Storage/FileStorage.php (hunk 3 above)
Comment (emulgeator) [reply src/Storage/FileStorage.php]: PHPstorm

File: src/Storage/DummyStorage.php [hunk 10]
```diff
@@ -1,175 +1,79 @@
...
-	 */
-	public function getConfigData() {
-		return array(
-			'storageType'      => StorageFactory::TYPE_DUMMY,
-			'debuggerDisabled' => $this->debuggerDisabled,
-		);
-	}
-}
\ No newline at end of file
+class DummyStorage extends StorageAbstract
```
[the remark this answers is not in the document]
Comment (emulgeator) [reply src/Storage/DummyStorage.php]: But it's a dummy storage.
If you tell it wont store anything. Though you have the possibility to make it store the data internally for debugging reasons it's still a dummy storage as there's nothing beneath technically it's a special global variable nothing more

File: src/Debug/DataHandlerRegistry.php [hunk 11]
```diff
@@ -0,0 +1,123 @@
+<?php
+declare(strict_types = 1);
```
Comment (szeber) [inline src/Debug/DataHandlerRegistry.php]: Didn't we agree to not put spaces around this? Since that's in the draft PSR.

File: src/Debug/DefaultDataHandler.php [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        $this->generalItems[] = $item;
+    }
+
+    public function getCurlRequestItems(): array
+    {
+        return $this->curlRequestItems;
+    }
+
+    public function getErrorItems(): array
```
Comment (szeber) [inline src/Debug/DefaultDataHandler.php]: Why did you remove these docblocks?

File: src/Debug/DataHandlerRegistry.php (hunk 11 above)
Comment (emulgeator) [reply src/Debug/DataHandlerRegistry.php]: You are the one who added this rule to the PHP CS fixer ;)

File: src/Debug/DefaultDataHandler.php (hunk 12 above)
Comment (emulgeator) [reply src/Debug/DefaultDataHandler.php]: They are defined in the the interface properly. Makes no sense to duplicate them and we agreed on not using @inheritdoc annotation
